libs/datamodel/core/src/transform/ast_to_dml/db/attributes.rs

```diff
@@ -50,55 +397,298 @@ pub(super) fn model_id(id_args: &mut Arguments<'_>, model_id: ast::ModelId, ctx:
+                        fields = suggested_fields.join(", ")
+                    )
+                } else {
+                    String::new()
+                };
+
+                ctx.push_error(DatamodelError::new_model_validation_error(
+                    &format!(
+                        "The {prefix}index definition refers to the relation fields {the_fields}. Index definitions must reference only scalar fields.{suggestion}",
+                        prefix = if index_data.is_unique { "unique " } else { "" },
+                        the_fields = relation_fields.iter().map(|(f, _)| &f.name.name).join(", "),
+                        suggestion = suggestion
+                    ),
+                    &ctx.db.ast[model_id].name.name,
+                    args.span(),
+                ));
+            }
+        }
+    }
+}
+
+pub(super) fn visit_map<'ast>(map_args: &mut Arguments<'ast>, mapped_name: MappedName, ctx: &mut Context<'ast>) {
+    match map_args.default_arg("name").map(|value| value.as_str()) {
+        Ok(Ok(name)) => {
+            ctx.db.names.mapped_names.insert(mapped_name, name);
+        }
+        Err(err) => ctx.push_error(err), // not flattened for error handing legacy reasons
+        Ok(Err(err)) => ctx.push_error(map_args.new_attribute_validation_error(&err.to_string())),
+    }
+}
+
+/// @relation validation for relation fields.
+fn visit_relation<'ast>(
+    relation_args: &mut Arguments<'ast>,
+    model_id: ast::ModelId,
+    relation_field: &mut RelationField<'ast>,
+    ctx: &mut Context<'ast>,
+) {
+    if let Some(fields) = relation_args.optional_arg("fields") {
+        let fields = match resolve_field_array(&fields, relation_args.span(), model_id, ctx) {
+            Ok(fields) => fields,
+            Err(FieldResolutionError::AlreadyDealtWith) => Vec::new(),
+            Err(FieldResolutionError::ProblematicFields {
+                unknown_fields: unresolvable_fields,
+                relation_fields,
+            }) => {
+                if !unresolvable_fields.is_empty() {
+                    ctx.push_error(DatamodelError::new_validation_error(&format!("The argument fields must refer only to existing fields. The following fields do not exist in this model: {}", unresolvable_fields.join(", ")), fields.span()))
+                }
+
+                if !relation_fields.is_empty() {
+                    ctx.push_error(DatamodelError::new_validation_error(&format!("The argument fields must refer only to scalar fields. But it is referencing the following relation fields: {}", relation_fields.iter().map(|(f, _)| &f.name.name).join(", ")), fields.span()));
+                }
+
+                Vec::new()
+            }
+        };
+
+        relation_field.fields = Some(fields);
+    }
+
+    if let Some(references) = relation_args.optional_arg("references") {
+        let references = match references.as_constant_array() {
```

Comment:
Aaah good catch. Before I made attribute validation separate from type resolution, the referenced model fields couldn't have been resolved there (because the model could be after the current one), but now it should work. I'll try.